### PR TITLE
Simple support for nftables

### PR DESCRIPTION
**Pull request by yujincheng08**
Thanks @amnore's previous work.

I follow that work but with a simpler implementation. Tested with OpenWrt 22.03.0-rc5.

### CODE REVIEW

File: etc/smartdns/smartdns.conf [hunk 1]
```diff
@@ -206,6 +206,8 @@ log-level info
 # ipset /www.example.com/block, set ipset with ipset name of block 
 # ipset /www.example.com/-, ignore this domain
 
+# nftset [no]
```
Comment (pymumu) [inline etc/smartdns/smartdns.conf:209]: Consider adding a configuration similar to the following:

```shell
nftset /domain/[setname]
ntfset-timeout [yes|no]
```

File: package/openwrt/files/etc/init.d/smartdns [hunk 2]
```diff
@@ -326,6 +326,9 @@ load_service()
 	config_get log_file "$section" "log_file" ""
 	[ -z "$log_file" ] || conf_append "log-file" "$log_file"
 
+	config_get nftset "$section" "nftset" "0"
```
Comment (pymumu) [inline package/openwrt/files/etc/init.d/smartdns:329]: no need to add this options, when the nftset configuration item exists, the feature can be automatically enabled

File: src/dns_server.c [hunk 3]
```diff
@@ -1306,9 +1306,20 @@ static int _dns_server_setup_ipset_packet(struct dns_server_post_context *contex
 				}
 
 				/* add IPV4 to ipset */
-				ipset_add(rule->ipsetname, addr, DNS_RR_A_LEN, request->ip_ttl * 2);
-				tlog(TLOG_DEBUG, "IPSET-MATCH: domain: %s, ipset: %s, IP: %d.%d.%d.%d", request->domain,
-					 rule->ipsetname, addr[0], addr[1], addr[2], addr[3]);
+#ifdef WITH_NFTSET
```
Comment (pymumu) [inline src/dns_server.c:1309]: we can support both ipset and nftset at same time.

File: src/util.c [hunk 4]
```diff
@@ -632,6 +640,66 @@ int ipset_del(const char *ipsetname, const unsigned char addr[], int addr_len)
...
+						   const char *flags)
+{
+	char *cmd_buf = NULL;
+
+	if (_nftset_init() != 0) {
+		return -1;
+	}
+
+	char addr_str[INET6_ADDRSTRLEN];
+	if (!inet_ntop(af, addr, addr_str, INET6_ADDRSTRLEN)) {
```
Comment (pymumu) [inline src/util.c:668]: memory leak here for nft_ctx_new

File: src/dns_conf.c [hunk 5]
```diff
@@ -650,6 +653,15 @@ static int _conf_domain_rule_ipset(char *domain, const char *ipsetname)
 			continue;
 		}
 
+		// replace '/' to ' ' so as to support nftset
+		char* i = tok;
```
Comment (pymumu) [inline src/dns_conf.c:657]: This code should not be used. This processing may cause the ipset function to fail. When adding ipset later, you can get the IP and ipset name.

File: src/util.c (hunk 4 above)
Comment (yujincheng08) [reply src/util.c:668]: it's static so we dont hv to release it

File: etc/smartdns/smartdns.conf (hunk 1 above)
Comment (yujincheng08) [reply etc/smartdns/smartdns.conf:209]: for nft, we hv to specify three things: table type, table name and set name. should we consider a better config grammar for it?

File: src/dns_server.c (hunk 3 above)
Comment (yujincheng08) [reply src/dns_server.c:1309]: make sense. my thought was to follow some openwrt packages, e.g. miniupnd, whcih only support one backend: user can choose to install miniupnpd-iptables or miniupnpd-nftables. but anyway it make senses to support  both backends.

File: etc/smartdns/smartdns.conf (hunk 1 above)
Comment (pymumu) [reply etc/smartdns/smartdns.conf:209]: like this?
nftset /example.com/4#ip#test#setv4
nftset /example.com/6#ip#test#setv6

Comment (yujincheng08) [reply etc/smartdns/smartdns.conf:209]: lgtm, will follow this.

**Comment by yujincheng08** [comment]
@pymumu 改好了，再看看

**Pull request closed without merging**